compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -779,7 +779,28 @@ trait Implicits:
           SelectionProto(name, memberProto, compat, privateOK = false)
         case tp => tp
       }
-      try inferImplicit(adjust(to), from, from.span)
+
+      def isOldStyleFunctionConversion(tpe: Type): Boolean =
+        tpe match {
+          case PolyType(_, resType) =>
+            isOldStyleFunctionConversion(resType)
+          case _ => tpe.derivesFrom(defn.FunctionClass(1)) && !tpe.derivesFrom(defn.ConversionClass) && !tpe.derivesFrom(defn.SubTypeClass)
```

Comment:
It seems the code does not handle the following cases:

```Scala
implicit def implicitLength[A](implicit a: A): String => Int = _.length
implicit def implicitLength(implicit a: Dummy): String => Int = _.length
```

The following snippet might be useful:

```Scala
tpe match {
  case tp: MethodOrPoly => isOldStyleFunctionConversion(tp.finalResultType)
}
```

It would be good to add a test case:

``` Scala
def foo()(implicit x: String => Int) = ???
implicit val f: String => Int = _.size
foo()  // no warning
```

Reply:
This is indeed quite surprising but in both cases below
```
implicit def implicitLength[A](implicit a: A): String => Int = _.length
implicit def implicitLength(implicit a: Dummy): String => Int = _.length
```
`implicitLength` does NOT work as an implicit conversion but at least this behaviour is consistent between dotty and scala 2.13

Regarding
```
def foo()(implicit x: String => Int) = ???
implicit val f: String => Int = _.size
foo()  // no warning
```
which use case exactly is it going to check? If it's just to verify that no waring is raised if a function is passed as an implicit parameter but it's not used as an implicit conversion then this is already covered here
```
object Test8 {
  implicit def a2int[A](a: A)(implicit ev: A => Int): Int = ev(a) // ok
}
```

Reply:
> implicitLength does NOT work as an implicit conversion but at least this behaviour is consistent between dotty and scala 2.13

Thanks, that is good to know.

> which use case exactly is it going to check?

It's intended to check that an implicit function is used as an implicit argument instead of implicit conversion.

Reply:
ok, so this case is already covered as stated above

Reply:
A minor issue with the coverage: in `Test8`, `ev` is used _explicitly_. The test I suggested uses it _implicitly_.

Reply:
I would say in your example `ev` is finally not used at all. But I added that case anyway.

Reply:
It turns out that `case tp: MethodOrPoly => isOldStyleFunctionConversion(tp.finalResultType)` is wrong and it caused bootstrapped tests to fail. I reverted this change and added a new test case for that.

Reply:
It would be good to figure out the cause. What about the following test case:

```Scala
  implicit def mySeq2seq[A]: Seq[A] = ???
  def foo(implicit ev: Seq[Int]): Unit = ???
  foo
```

Reply:
It looks like `tp.finalResultType` simply goes too far. In the test case
```
object Tes14 {
  case class MySeq[A](underlying: Seq[A])

  implicit def mySeq2seq[A](mySeq: MySeq[A]): Seq[A] = mySeq.underlying
  val s: Seq[Int] = MySeq(Seq(1, 2, 3)) // ok
}
```
`mySeq2seq` is a method so it shouldn't be reported as deprecated conversion and in this case `case PolyType(_, resType)`, `resType` is a MethodType while `tp.finalResultType` would return the final return type of that method, which is `Seq[A]` and which happens to be a subtype of `Function1` so `isOldStyleFunctionConversion` finally returns true.

The test case you're suggesting now succeeds for both implementation variants.

Reply:
Thanks, it looks good to me now.